add ZKReq.html (#13750)
taikoxyz/taiko-mono: packages/protocol/docs/ZKReq.md

A valid transaction (defined in the Ethereum Yellow Paper):

- Has a valid transaction signature.
- Has a valid transaction nonce (equivalent to the sender account's current nonce).
- Has no contract code deployed on the sender account (see EIP-3607 by Feist et al. [2021]).
- Has a gas limit no smaller than the intrinsic gas, **`g0`**, used by the transaction; and the sender account balance contains at least the cost, **`v0`**, required in up-front payment.

In contrast, the previous implementation allowed for invalid blocks that were skipped. The new protocol maps all (valid and invalid) **`txLists`** into L2 blocks.

## Data Consistency

Note that `txList` is not directly know to L1 contracts, therefore, ZKP shall further prove that the chosen `txList` is a slice of the given blob data (from `meta.txListByteStart` to `meta.txListByteEnd`) and the blob’s hash or polynomial commitment (PC) is `meta.blobHash` — not the `txList`’s hash or PC). ( @todo/question: blobHash is not part of the contracts anymore.)

## Anchor Transactions

Every L2 block has exactly one anchor function call as its first transaction.

> ZKP must prove that *TaikoL2.anchor(...)* is the first transaction in the block, with the correct input parameters and gas limit, signed by the gold-touch address, and did not throw.

Author: It's still a regular transaction based on what I know.

- The actual value of the input parameters must be consistent with the values used to hash the ZKP instance (see [**LibProving**](https://github.com/taikoxyz/taiko-mono/blob/80da7124a2e7340f65aa88561f51b69a79404577/packages/protocol/contracts/L1/libs/LibProving.sol#L118)).

Note that the anchor transaction emits a `Anchored` event that may help ZKP to verify block variables. See below.

### Anchor Signature

ZKP shall also check the signature of the anchor transaction:

- The signer must be TaikoL2.GOLDEN_TOUCH_ADDRESS.
- The signature must use `1` or `2` as the `K` value. See *TaikoL2Signer.sol*.
- `K=1` must be used unless [….forgot]

[EDIT: this will become a new transaction type]

# The *BlockMetadata*

This struct represents the metadata of a block. The data will be hashed and be part of the public input. So anything within this struct, will be hashed and put into ZKP `instance`.

- `id`: Represents the block height in L2.
- `timestamp`: The block timestamp in L2.
- `l1Height`: The actual block height in L1.
- `l1Hash`: The actual block hash in L1.
- `mixHash`: Salted random number to accommodate multiple L2 blocks fitting into one L1 block.

> - `depositsRoot`: Matches the withdrawalsRoot in the L2 header.

Reviewer: ```suggestion
- `depositsRoot`: Hash of the list of deposits.
```

- `block.gaslimit`: ZKP must verify this value must equal `meta.gasLimit`.
- `block.gasUsed`: ZKP must verify this value must be smaller than `meta.gasLimit`.
- `block.number`: this must be checked against the block header and `meta.id`.
- `block.prevrandao`: this must be checked against the `mixHash` field in the L2 block header and `meta.mixHash`.
- `block.timestamp`: this must be checked against the `timestamp` field in the L2 block header and `meta.proposedAt`.

## Block Header

The block header is no longer presented to L1 contracts; therefore, ZKP must verify the integrity of the block header and ensure the block header can hash into the same value as `evidence.blockHash`, which is part of the ZK instance input.

In addition, ZKP must also prove the following:

- `parentHash` must be the same as `evidence.parentHash`.
- `ommersHash` must be `0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347`.
- `beneficiary` must be `meta.beneficiary` (duplicated, as stated above).
- `logsBloom` must be a bytes32[8] with all zeros.
- `difficulty` must be `meta.mixHash`.
- `height` must be `meta.id`.
- `gasLimit` == `meta.gasLimit`.

> - `gasUsed` ≤ `gasLimit` (we are assuming the EVM proof will verify that gasLimit has the right value), and `gasUsed` == `evidence.gasUsed`

Reviewer: ```suggestion
- `gasUsed` ≤ `gasLimit` (we are assuming the ZKP will verify that gasLimit has the right value), and `gasUsed` == `evidence.gasUsed`
```

- `timestamp` == `meta.proposedAt`.
- `extraData` == "".
- `mixHash` == `meta.mixHash`.
- `nonce` == 0.

> - `baseFeePerGas` == 0.

Reviewer: ```suggestion
- `baseFeePerGas` == the calculated EIP-1559 style base fee.
```

> - `withdrawalsRoot` == `0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421`  according to [https://github.com/ethereum/go-ethereum/blob/master/core/types/hashes.go#L41](https://github.com/ethereum/go-ethereum/blob/master/core/types/hashes.go#L41)

Reviewer: ```suggestion
- `withdrawalsRoot` == The withdrawals MPT correspoding to the deposits list.
```

Note that some of the above header field checks are duplicates of checks from the Block Variable section.

## Signal Storage Root

ZKP also needs to prove cross chain signal service’s storage roots have the right value.

> - **For L2 Signal Service**:  the storage root is the second parameter in the anchor transaction. ZKP shall verify that the storage root of the L2 Signal Service address indeed has this given value by using zkEVM storage proof.

Reviewer: ```suggestion
- **For L2 Signal Service**:  the storage root is the second parameter in the anchor transaction. The ZKP shall verify that the storage root of the L2 Signal Service address has this given value by using a standard zkEVM storage proof.
```

> - **For L1 Signal Service**: the storage root verification shall be done in circuits by using a Merkle proof as witness against `meta.l1Hash`. The `l1SignalRoot` value is not directly provided by L1 smart contracts as it is not available to EVM and must be queried by L2 client from a L1 node.

Reviewer: ```suggestion
- **For L1 Signal Service**: the storage root verification will be done in the circuits by using a Merkle proof against the `meta.l1Hash`, the `l1SignalRoot` value is not directly available in the L1 smart contracts. This MPT proof must be queried from an L1 node by the L2 client.
```

## Data cross-verification chart

```mermaid
graph LR
classDef default stroke-width:4px;

m_id --- h_height --- v_block_number;
m_gas_limit --- h_gas_limit --- v_block_gaslimit;
m_timestamp --- h_timestamp  --- v_block_timestamp;
m_h1_height --- a_h1_height;
m_h1_hash --- a_h1_hash;
m_mix_hash --- h_mix_hash --- v_block_prevrando;
tx_list -.->|keccak| m_txlist_hash;
m_beneficiary --- h_beneficiary;
h_parent_hash --- v_blockhash_1 & e_parent_hash;
empty_list -.->|keccak| h_ommers_hash;
empty_list -.-> h_logs_bloom;
zero -.-> h_difficulty;
zero -.-> h_nonce;
empty_string -.-> h_extra_data;
l2_treasure -.-> m_treasure;
processed_deposits_data -.-> m_deposits;

v_block_chainid -.-> dot1;
v_blockhash_others -.-> dot1 -.->|keccak| s_public_input_hash;

v_block_basefee -.-> h_basefee;

v_block_gaslimit -.-> dot2;
v_block_timestamp -.-> dot2;
s_parent_timestamp -.-> dot2;
s_gas_excess -.-> dot2 ---|calcBasefee| v_block_basefee;

> processed_deposits -.->|keccak| m_deposits_root --- h_withdrawals_root;

Reviewer: It's not the same value, but it's deduced from it so needs to be a dotted line?

m_cache_txlist_info(cacheTxListInfo)
m_treasure(treasure)
m_beneficiary(beneficiary)
m_deposits_root(depositsRoot)
m_deposits(depositsProcessed)
end

subgraph BlockHeader
h_height(height)
h_gas_limit(gasLimit)
h_gas_used(gasUsed)
h_timestamp(timestamp)
h_mix_hash(mixHash)
h_beneficiary(beneficiary)
h_parent_hash(parentHash)
h_ommers_hash(ommersHash)
h_state_root(stateRoot)
h_transactions_root(transactionsRoot)
h_receipts_root(receiptsRoot)
h_logs_bloom(logsBloom)
h_difficulty(difficulty)
h_extra_data(extraData)
h_nonce(nonce)
h_basefee(basefee)
h_withdrawals_root(withdrawalsRoot)
end

subgraph GlobalVariables
v_block_number(block.number)
v_block_gaslimit(block.gaslimit)
v_block_timestamp(block.timestamp)
v_block_prevrando(block.prevrando)
v_blockhash_1("blockhash(1)")
v_blockhash_others("blockhash(2..256)")
v_block_chainid("block.chainid")
v_block_basefee("block.basefee")
dot1((" "))
dot2((" "))
end

subgraph Anchor
a_h1_height(h1Height)
a_h1_hash(h1Hash)
a_h1_signal_root[h1SignalRoot]
a_parent_gas_used[parentGasUsed]
end

subgraph L1Storage
b_signal_root[signalRoot]
b_l1_taiko_addr[taikoL1Address]
b_l1_signal_service_addr[L1 signalServiceAddress]

> b_l2_signal_service_addr[L2 signalServiceAddress]

Reviewer: I guess it depends a bit what you see under `L1Storage`, but I wouldn't put the `L2 signalServiceAddress` here.

### Block Metadata

This struct represents a proposed L2 block. The data will be hashed and be part of the public input to circuits.

```solidity
struct BlockMetadata {
  uint64 id;
  uint64 timestamp;
  uint64 l1Height;
  bytes32 l1Hash;
  bytes32 mixHash;
  bytes32 depositsRoot;
  bytes32 txListHash;
  uint24 txListByteStart;
  uint24 txListByteEnd;
  uint32 gasLimit;
  address beneficiary;
  uint8 cacheTxListInfo;
  address treasure;
  TaikoData.EthDeposit[] depositsProcessed;
}
```

- `id`: Represents the block height in L2.
- `timestamp`: The block timestamp in L2.
- `l1Height`: The actual block height in L1.
- `l1Hash`: The actual block hash in L1.
- `mixHash`: Salted random number to accommodate multiple L2 blocks fitting into one L1 block.
- `depositsRoot`: Hash of the list of deposits.
- `txListHash`: Hash of the transaction list in L2.
- `txListByteStart`: Byte start of the transaction list in L2.
- `txListByteEnd`: Byte end of the transaction list in L2.
- `gasLimit`: Gas limit for the L2 block.
- `beneficiary`: The address of the beneficiary in L2.

> - `cacheTxListInfo`: Indicates whether the transaction list info should be cached or not.

Author: This `cacheTxListInfo` indicates whether or not to cache the current `txList` in the transaction if the  `txList` is not empty. 

To use an cached txList (regardless if it is a blob or a calldata), the current block shall provide an empty `txList` to indicate to reuse the cached txList data.

> - `cacheTxListInfo`: Indicates whether the transaction list info should be cached or not.

Reviewer: Sure, but that has no impact on the circuit which always just gets the `txListHash` (or blob commitment in the future) as the input no?

> - `cacheTxListInfo`: Indicates whether the transaction list info should be cached or not.

Reviewer: If the purpose of the value is to just let the circuits know if a tx list is "cached" or not, then yes it should be removed. That information is not useful for the circuits.

But if the meaning of the variable is to decode the data from the tx list or blob data then that would be useful to put in the circuit.

> - `cacheTxListInfo`: Indicates whether the transaction list info should be cached or not.

Author: Please see this PR: https://github.com/taikoxyz/taiko-mono/pull/13793/files

- `blockhash`: EVM allows access to the most recent 256 block hashes. All these hashes are available inside the plonk lookup table. ZKP must prove that the lookup table is consistent with L2's historical values.
  - `blockhash(block.number - 1)`, has the same value as in the block header and is also the same value as the parent block's hash on L2.
  - The other 255 hashes, `blockhash(block.number - 256)` to `blockhash(block.number - 2)` are checked in the anchor transaction to simplify circuits. Therefore, as long as the anchor transaction is zk-proven, these 255 ancestor hashes are proven indirectly.
- `block.basefee`: verified to be the correct value in the anchor transaction.
- `block.chainid`: this field is also checked by the anchor transaction, so no extra ZKP circuits are required.
- `block.coinbase`: ZKP must verify the value must be the same as `meta.beneficiary`. Again, the metadata hash is part of the ZK instance.
- `block.difficulty`: this is now the same as `block.prevrandao`, so we only check `block.prevrandao`.
- `block.gaslimit`: ZKP must verify this value must equal `meta.gasLimit`.

> - `block.gasUsed`: ZKP must verify this value must be smaller than `meta.gasLimit`.

Reviewer: `gasUsed` is not accessible in the EVM so should be removed here.

- `block.number`: this must be checked against the block header and `meta.id`.
- `block.prevrandao`: this must be checked against the `mixHash` field in the L2 block header and `meta.mixHash`.
- `block.timestamp`: this must be checked against the `timestamp` field in the L2 block header and `meta.proposedAt`.

### Block Header

> The block header is no longer presented to L1 contracts; therefore, ZKP must verify the integrity of the block header and ensure the block header can hash into the same value as `evidence.blockHash`, which is part of the ZK instance input.

Reviewer: ```suggestion
Not all block header data is available in the L1 contracts; therefore, the ZKP must verify the integrity of the block header and ensure the block header hashes into the same value as `evidence.blockHash`, which is part of the ZK instance input.
```

```solidity
struct BlockHeader {
  bytes32 parentHash;
  bytes32 ommersHash;
  address beneficiary;
  bytes32 stateRoot;
  bytes32 transactionsRoot;
  bytes32 receiptsRoot;
  bytes32[8] logsBloom;
  uint256 difficulty;
  uint128 height;
  uint64 gasLimit;
  uint64 gasUsed;
  uint64 timestamp;
  bytes extraData;
  bytes32 mixHash;
  uint64 nonce;
  uint256 baseFeePerGas;
  bytes32 withdrawalsRoot;
}
```

In addition, ZKP must also prove the following:

- `parentHash` must be the same as `evidence.parentHash`.
- `ommersHash` must be the keccak256 of `[]`,or `0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347`.
- `beneficiary` must be `meta.beneficiary` (duplicated, as stated above).
- `logsBloom` must be a `bytes32[8]` with all zeros.

> - `difficulty` must be `meta.mixHash`.

Reviewer: ```suggestion
- `difficulty` == 0.
```

As @ggkitsas has found, this should not be the `mixHash`, the `DIFFICULTY` opcode just returns the `mixHash/prevRandao`. Its value in the block header is simply `0`: https://blog.ethereum.org/2021/11/29/how-the-merge-impacts-app-layer. In the diagram below it is correct.

- `height` must be `meta.id`.
- `gasLimit` == `meta.gasLimit`.

> - `gasUsed` ≤ `gasLimit` (we are assuming the ZKP will verify that gasLimit has the right value), and `gasUsed` == `evidence.gasUsed`

Reviewer: ```suggestion
- `gasUsed` must be the sum of all gas used by all valid transactions in the block.
```

- `timestamp` == `meta.proposedAt`.
- `extraData` == "".
- `mixHash` == `meta.mixHash`.
- `nonce` == 0.
- `baseFeePerGas` == the calculated EIP-1559 style base fee.

> - `withdrawalsRoot` == The withdrawals MPT correspoding to the deposits list.

Reviewer: ```suggestion
- `withdrawalsRoot` == The withdrawals MPT root corresponding to the deposits list.
```

> Note that some of the above header field checks are duplicates of checks from the Global Variable section.

Reviewer: ```suggestion
Note that some of the header field checks above are duplicates of checks done in the Global Variable section.
```

### Signal Storage

> ZKP also needs to prove cross chain signal service’s storage roots have the right values.

Reviewer: ```suggestion
The ZKP also needs to prove that the cross chain signal service’s storage roots have the correct values.
```

> - **For L2 Signal Service**: the storage root is the second parameter in the anchor transaction. ZKP shall verify that the storage root of the L2 Signal Service address indeed has this given value by using zkEVM storage proof.

Reviewer: ```suggestion
- **For L2 Signal Service**: the L1 storage root of the signal service is the second parameter in the anchor transaction. The ZKP shall verify that the storage root of the L1 Signal Service address has the given value by using an MPT proof against the state root stored in `meta.l1Hash` for the `l1SignalServiceAddress` account. This MPT proof must be queried by the L2 client from an L1 node .
```

> - **For L1 Signal Service**: the storage root verification will be done in the circuits by using a Merkle proof against the `meta.l1Hash`, the `l1SignalRoot` value is not directly available in the L1 smart contracts. This MPT proof must be queried from an L1 node by the L2 client.

Reviewer: ```suggestion
- **For L1 Signal Service**: the L2 storage root verification will be done in the circuits by using an MPT proof against the post block state root for the `l2SignalServiceAddress` account.
```
